vortex-datafusion/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+//! Connectors to enable DataFusion to read Vortex data.
+
+use std::any::Any;
+use std::fmt::Formatter;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use arrow_array::RecordBatch;
+use arrow_schema::SchemaRef;
+use arrow_schema::{DataType, Field, Schema};
+use async_trait::async_trait;
+use datafusion::datasource::TableProvider;
+use datafusion::execution::context::SessionState;
+use datafusion::execution::{RecordBatchStream, SendableRecordBatchStream, TaskContext};
+use datafusion_common::{exec_datafusion_err, exec_err, DataFusionError, Result as DFResult};
+use datafusion_expr::{Expr, TableProviderFilterPushDown, TableType};
+use datafusion_physical_expr::EquivalenceProperties;
+use datafusion_physical_plan::{
+    DisplayAs, DisplayFormatType, ExecutionMode, ExecutionPlan, Partitioning, PlanProperties,
+};
+use futures::{Stream, StreamExt};
+use pin_project::pin_project;
+use vortex::array::chunked::ChunkedArray;
+use vortex::array::r#struct::StructArray;
+use vortex::compute::as_arrow::AsArrowArray;
+use vortex::{Array, ArrayDType, Flattened, IntoArray};
+use vortex_dtype::DType;
+use vortex_error::{vortex_bail, VortexResult};
+
+/// A [`TableProvider`] that exposes an existing Vortex Array to the DataFusion SQL engine.
+///
+/// Only arrays that have a top-level [struct type](vortex_dtype::StructDType) can be exposed as
+/// a table to DataFusion.
+#[derive(Debug, Clone)]
+pub(crate) struct VortexInMemoryTableProvider {
+    array: Array,
+    schema_ref: SchemaRef,
+}
+
+impl VortexInMemoryTableProvider {
+    /// Build a new table provider from an existing [struct type](vortex_dtype::StructDType) array.
+    pub fn try_new(array: Array) -> VortexResult<Self> {
+        if !matches!(array.dtype(), DType::Struct(_, _)) {
+            vortex_bail!(InvalidArgument: "only DType::Struct arrays can produce a table provider");
+        }
+
+        let arrow_schema = Schema::try_from(array.dtype())?;
```

Comment:
We actually removed this conversion since it is kind of lossy. You can't tell just from a Vortex DType the exact Arrow type, e.g. is it a String or a LargeString?

What we wanted to add was `array.arrow_dtype()` possibly related to the AsArrow compute function, or add an AsArrowDType compute function or something....

Reply:
I think VarBin and VarBinView are the only places where this is ambiguous, and for those you can introspect the encoding details. I think it makes most sense to just implement this as a method on `Flattened`...does that sound right?

Reply:
This will happen more often - timestamp/date/time will be likely different. Similarly for list/fixedsizedlist/largelist/listview